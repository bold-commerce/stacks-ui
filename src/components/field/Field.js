import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';

import './Field.css';

function Field(props) {
  const {
    children, className, disabled, helpText, htmlFor, id, messageId, label, messageText, messageType, ariaLive, ...otherProps
  } = props;

  const classNames = cn([
    'stx-field',
    { 'stx-field--alert': messageType === 'alert' || messageType === 'error' },
    { 'stx-field--warning': messageType === 'warning' },
    { 'stx-field--success': messageType === 'success' },
    { 'stx-field--is-disabled': disabled },
    className,
  ]);

  const renderMessageText = () => {
    return messageText && <div className='stx-field__message' aria-live={ariaLive} id={messageId}>{ messageText }</div>;
  };

  const renderHelpText = () => {
    return helpText && <div className='stx-field__help-text'>{ helpText }</div>;
  };

  return (
    <div
      {...otherProps}
      id={id}
      className={classNames}
    >
      { label && <label htmlFor={htmlFor} className='stx-field__label'> { label } </label> }

      { children }

      { renderMessageText() }

      { renderHelpText() }
    </div>
  );
};

Field.defaultProps = {};

Field.propTypes = {
  /** Text for the label element in for the field */
  label: PropTypes.node,

  /** Text for additional text that appears below the form control */
  helpText: PropTypes.node,

  /** The 'for' attribute on the label to link it to the form control via ID */
  htmlFor: PropTypes.string,

  /** The ID attribute for the field component */
  id: PropTypes.string,

  /** Optional classNames for the field component */
  className: PropTypes.string,

  /** Whether or not the field is disabled */
  disabled: PropTypes.bool,

  /** The type of message to show */
  messageType: PropTypes.oneOf(['success', 'warning', 'alert', '']),

  /** The aria-live attribute value that will be set on the message if one is given */
  ariaLive: PropTypes.oneOf(['polite', 'assertive']),

  /** The text to display for the message */
  messageText: PropTypes.string,

  /** A unique ID for the message element */
  messageId: PropTypes.string,

  /** Child elements */
  children: PropTypes.node,
};

export default Field;
