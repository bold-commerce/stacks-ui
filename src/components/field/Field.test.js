import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Field from './Field';

describe('Field', () => {
  test('Field renders with basic props', () => {
    const output = mount(<Field
      className='the-class'
      data-additional-prop='additional-prop'
      messageText='Some message text'
      messageType='alert'
      helpText='Some help text'
      ariaLive='assertive'
    >
      <div>
        {'There are children'}
      </div>
    </Field>);

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Field render disabled when disabled is true', () => {
    const output = shallow(<Field disabled />);
    expect(toJson(output)).toMatchSnapshot();
  });
});
