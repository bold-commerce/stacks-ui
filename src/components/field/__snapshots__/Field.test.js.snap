// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Field render disabled when disabled is true 1`] = `
<div
  className="stx-field stx-field--is-disabled"
/>
`;

exports[`Field Field renders with basic props 1`] = `
<Field
  ariaLive="assertive"
  className="the-class"
  data-additional-prop="additional-prop"
  helpText="Some help text"
  messageText="Some message text"
  messageType="alert"
>
  <div
    className="stx-field stx-field--alert the-class"
    data-additional-prop="additional-prop"
  >
    <div>
      There are children
    </div>
    <div
      aria-live="assertive"
      className="stx-field__message"
    >
      Some message text
    </div>
    <div
      className="stx-field__help-text"
    >
      Some help text
    </div>
  </div>
</Field>
`;
